src/main/java/com/google/api/codegen/config/DiscoveryField.java

```diff
@@ -250,4 +245,185 @@ public Oneof getOneof() {
+        throw new IllegalArgumentException(
+            "Tried to assign value for unsupported Schema type "
+                + schema.type()
+                + ", format "
+                + schema.format()
+                + "; value "
+                + value);
+    }
+    throw new IllegalArgumentException(
+        "Could not assign value '"
+            + value
+            + "' to type "
+            + schema.type()
+            + ", format "
+            + schema.format());
+  }
+
+  @Override
+  public List<? extends FieldModel> getFields() {
+    return properties;
+  }
+
+  @Override
+  // Schemas are immutable, so this is just the identity function.
+  public TypeModel makeOptional() {
+    return this;
+  }
+
+  @Override
+  public String getPrimitiveTypeName() {
+    Preconditions.checkArgument(isPrimitiveType());
+    switch (schema.type()) {
+      case INTEGER:
+        switch (schema.format()) {
+          case UINT32:
+            return "uint32";
+          default:
+            return "int32";
+        }
+      case NUMBER:
+        switch (schema.format()) {
+          case FLOAT:
+            return "float";
+          case DOUBLE:
+          default:
+            return "double";
+        }
+      case BOOLEAN:
+        return "boolean";
+      case STRING:
+        if (schema.format() == null) {
+          return "string";
+        }
+        switch (schema.format()) {
+          case BYTE:
+            return "bytes";
+          case INT64:
+            return "sint64";
+          case UINT64:
+            return "uint64";
+          default:
+            return "string";
+        }
+      default:
+        return null;
+    }
+  }
+
+  private boolean isPrimitiveType() {
+    return schema.type().equals(Type.BOOLEAN)
+        || schema.type().equals(Type.INTEGER)
+        || schema.type().equals(Type.NUMBER)
+        || schema.type().equals(Type.STRING);
+  }
+
+  @Override
+  public boolean isBooleanType() {
+    return schema.type().equals(Type.BOOLEAN);
+  }
+
+  @Override
+  public boolean isStringType() {
+    return schema.type().equals(Type.STRING) && schema.format() == null;
+  }
+
+  @Override
+  public boolean isFloatType() {
+    return schema.type().equals(Type.NUMBER) && schema.format().equals(Format.FLOAT);
+  }
+
+  @Override
+  public boolean isBytesType() {
+    return schema.type().equals(Type.STRING) && schema.format().equals(Format.BYTE);
+  }
+
+  @Override
+  public boolean isDoubleType() {
+    return schema.type().equals(Type.NUMBER) && schema.format().equals(Format.DOUBLE);
+  }
+
+  @Override
+  public String getTypeName() {
+    if (isPrimitiveType()) {
+      return getPrimitiveTypeName();
+    }
+    switch (schema.type()) {
+      case ARRAY:
+        return "list";
+      default:
+        return "message";
+    }
+  }
+
+  @Override
+  public DiscoveryField getType() {
+    return this;
+  }
+
+  @Override
+  public boolean isEmptyType() {
+    return false;
```

Comment:
The empty type is an actual type, here's an example:
https://github.com/googleapis/toolkit/blob/master/src/test/java/com/google/api/codegen/testsrc/discoveries/pubsub.v1.json#L26

Dunno if you're already dealing with it somewhere else.

Reply:
cool this has been updated to reflect Empty being a concrete type.